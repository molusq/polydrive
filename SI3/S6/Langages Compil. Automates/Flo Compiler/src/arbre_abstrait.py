import enum
import json
from typing import List, Optional, TypeAlias

JSON: TypeAlias = dict[str,
                       "JSON"] | list["JSON"] | str | int | float | bool | None


class Type(enum.Enum):
    INCONNU = enum.auto()
    ENTIER = enum.auto()
    BOOLEEN = enum.auto()
    FONCTION = enum.auto()
    VIDE = enum.auto()

    @staticmethod
    def from_str(s: str) -> "Type":
        if s == "entier":
            return Type.ENTIER
        if s == "booleen":
            return Type.BOOLEEN
        return Type.INCONNU


class AST:

    def to_json(self) -> JSON:
        print(self.__class__.__name__)
        raise NotImplementedError("to_json not implemented")

    def type(self) -> Type:
        return Type.INCONNU

    def __str__(self):
        return json.dumps(self.to_json(), indent=2)


class Programme(AST):

    def __init__(self, instructions: List[AST]):
        self.instructions = instructions

    def type(self) -> Type:
        raise NotImplementedError("Programme has no type")
        return Type.AUTRE

    def to_json(self) -> JSON:
        return {"instructions": [i.to_json() for i in self.instructions]}

    def addInstruction(self, instruction: AST):
        self.instructions.append(instruction)


class Operation(AST):

    def __init__(self, op: str, exp1: AST, exp2: AST):
        self.lhs = exp1
        self.op = op
        self.rhs = exp2

    def type(self) -> Type:
        if self.lhs.type() == self.rhs.type():
            if self.op in ["<", "<=", ">", ">=", "==", "!="]:
                return Type.BOOLEEN
            return self.lhs.type()
        print("op:", self.lhs.to_json(), self.rhs.to_json())
        return Type.INCONNU

    def to_json(self) -> JSON:
        return {
            "op": self.op,
            "lhs": self.lhs.to_json(),
            "rhs": self.rhs.to_json()
        }


class OperationUnaire(AST):

    def __init__(self, op: str, exp: AST):
        self.op = op
        self.exp = exp

    def type(self) -> Type:
        return self.exp.type()

    def to_json(self) -> JSON:
        return {"op": self.op, "exp": self.exp.to_json()}


class Entier(AST):

    def __init__(self, valeur: int):
        self.valeur = valeur

    def type(self) -> Type:
        return Type.ENTIER

    def __eq__(self, rhs: object) -> bool:
        if isinstance(rhs, Entier):
            return self.valeur == rhs.valeur
        return self.valeur == rhs

    def to_json(self) -> JSON:
        return {"entier": self.valeur}


class Booleen(AST):

    def __init__(self, valeur: bool):
        self.valeur = valeur

    def type(self) -> Type:
        return Type.BOOLEEN

    def __eq__(self, rhs: object) -> bool:
        if isinstance(rhs, Booleen):
            return self.valeur == rhs.valeur
        return self.valeur == rhs

    def to_json(self) -> JSON:
        return {"booleen": self.valeur}


class Identifiant(AST):

    def __init__(self, valeur: str):
        self.valeur = valeur

    def type(self) -> Type:
        return Type.AUTRE

    def __eq__(self, rhs: object) -> bool:
        if isinstance(rhs, Identifiant):
            return self.valeur == rhs.valeur
        return self.valeur == rhs

    def to_json(self) -> JSON:
        return {"identifiant": self.valeur}


class If(AST):

    def __init__(self, condition: AST, body: Programme,
                 orelse: Optional[Programme]):
        self.condition = condition
        self.body = body
        self.orelse = orelse or Programme([])

    def type(self) -> Type:
        return Type.AUTRE

    def add_elif(self, condition: AST, body: Programme):
        if self.orelse.instructions and isinstance(
                self.orelse.instructions[-1], If):
            self.orelse.instructions[-1].add_elif(condition, body)
        else:
            self.orelse.addInstruction(If(condition, body, None))

    def add_else(self, body: Programme):
        self.add_elif(Booleen(True), body)

    def to_json(self) -> JSON:
        return {
            "condition": self.condition.to_json(),
            "body": self.body.to_json(),
            "orelse": self.orelse.to_json()
        }


class While(AST):

    def __init__(self, condition: AST, body: Programme):
        self.condition = condition
        self.body = body

    def type(self) -> Type:
        return Type.AUTRE

    def to_json(self) -> JSON:
        return {
            "condition": self.condition.to_json(),
            "body": self.body.to_json()
        }


class Declaration(AST):

    def __init__(self, name: Identifiant, _type: Type, value: Optional[AST]):
        self.name = name
        self._type = _type
        self.value = value

    def type(self) -> Type:
        return self._type

    def to_json(self) -> JSON:
        return {
            "name": self.name.to_json(),
            "type": self.type().name,
            "value": self.value.to_json() if self.value else None
        }


class Assignment(AST):

    def __init__(self, name: Identifiant, value: AST):
        self.name = name
        self.value = value

    def type(self) -> Type:
        return Type.AUTRE

    def to_json(self) -> JSON:
        return {"name": self.name.to_json(), "value": self.value.to_json()}


class Return(AST):

    def __init__(self, value: AST):
        self.value = value

    def type(self) -> Type:
        return Type.AUTRE

    def to_json(self) -> JSON:
        return {"return_value": self.value.to_json()}


class Function(AST):

    def __init__(self, name: Identifiant, args: List[Declaration],
                 body: Programme, return_type: Type):
        self.name = name
        self.args = args
        self.body = body
        self.return_type = return_type
        self.stack_size = 0

    def type(self) -> Type:
        return Type.FONCTION

    def to_json(self) -> JSON:
        return {
            "name": self.name.to_json(),
            "args": [a.to_json() for a in self.args],
            "body": self.body.to_json(),
            "return_type": self.return_type.name,
            "stack_size": self.stack_size
        }


class FunctionArgs(AST):

    def __init__(self, args: List[AST]):
        self.args = args

    def type(self) -> Type:
        return Type.AUTRE

    def to_json(self) -> JSON:
        return list(map(lambda x: x.to_json(), self.args))

    def __getitem__(self, i: int):
        return self.args[i]

    def __iter__(self):
        return iter(self.args)

    def append(self, arg: AST):
        self.args.append(arg)

    def __len__(self):
        return len(self.args)


class AppelFonction(AST):

    def __init__(self, name: Identifiant, args: FunctionArgs):
        self.name = name
        self.args = args

    def type(self) -> Type:
        return Type.INCONNU

    def to_json(self) -> JSON:
        return {"function_name": self.name.valeur, "args": self.args.to_json()}
