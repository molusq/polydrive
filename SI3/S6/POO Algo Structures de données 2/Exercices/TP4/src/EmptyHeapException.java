
/**
 * A class exception for empty heap
 */
public class EmptyHeapException extends Exception {
	
	public EmptyHeapException() {
		super();
	}

}
