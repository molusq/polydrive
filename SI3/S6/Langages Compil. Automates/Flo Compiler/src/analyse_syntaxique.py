# type: ignore

from sly import Parser
from src.analyse_lexicale import FloLexer
import src.arbre_abstrait as arbre_abstrait


class FloParser(Parser):

    def error(self, token):
        '''
        Default error handling function.  This may be subclassed.
        '''
        if token:
            lineno = getattr(token, 'lineno', 0)
            if lineno:
                print(
                    f'sly: Syntax error at line {lineno}, token={token.type}\n'
                )
            else:
                print(f'sly: Syntax error, token={token.type}')
        else:
            print('sly: Parse error in input. EOF\n')
        print(list(self.symstack))

    # On récupère la liste des lexèmes de l'analyse lexicale
    tokens = FloLexer.tokens

    precedence = (
        ('left', 'ET'),
        ('left', 'OU'),
        ('right', 'NON'),
        ('right', 'EGAL'),
        ('left', 'DIFFERENT', 'INFERIEUR_OU_EGAL', 'SUPERIEUR_OU_EGAL',
         'INFERIEUR', 'SUPERIEUR'),
        ('left', '+', '-'),
        ('left', '*', '/', '%'),
        ('nonassoc', 'UMOINS'),
    )

    # Règles gramaticales et actions associées
    @_('instruction')
    def prog(self, p):
        return arbre_abstrait.Programme([p.instruction])

    @_('prog instruction')
    def prog(self, p):
        p.prog.addInstruction(p.instruction)
        return p.prog

    @_('expr ";"')
    def instruction(self, p):
        return p.expr

    @_('bool')
    def expr(self, p):
        return p[0]

    @_('BOOLEEN')
    def bool(self, p):
        return arbre_abstrait.Booleen(p[0])

    @_('bool ET bool')
    def bool(self, p):
        return arbre_abstrait.Operation('et', p[0], p[2])

    @_('bool OU bool')
    def bool(self, p):
        return arbre_abstrait.Operation('ou', p[0], p[2])

    @_('NON bool')
    def bool(self, p):
        return arbre_abstrait.OperationUnaire('non', p[1])

    @_('exprArith')
    def bool(self, p):
        return p[0]

    @_('exprArith "+" exprArith', 'exprArith "-" exprArith',
       'exprArith "*" exprArith', 'exprArith "/" exprArith',
       'exprArith "%" exprArith', 'exprArith INFERIEUR exprArith',
       'exprArith SUPERIEUR exprArith', 'exprArith EGAL exprArith',
       'exprArith DIFFERENT exprArith',
       'exprArith INFERIEUR_OU_EGAL exprArith',
       'exprArith SUPERIEUR_OU_EGAL exprArith')
    def exprArith(self, p):
        return arbre_abstrait.Operation(p[1], p[0], p[2])

    @_('"-" exprArith %prec UMOINS')
    def exprArith(self, p):
        return arbre_abstrait.OperationUnaire('-', p[1])

    @_('"(" bool ")"')
    def exprArith(self, p):
        return p[1]

    @_('facteur')
    def exprArith(self, p):
        return p[0]

    @_('identifiant "(" function_arg ")"')
    def facteur(self, p):
        return arbre_abstrait.AppelFonction(p.identifiant, p.function_arg)

    @_('identifiant "("  ")"')
    def facteur(self, p):
        return arbre_abstrait.AppelFonction(p.identifiant,
                                            arbre_abstrait.FunctionArgs([]))

    @_('ENTIER')
    def facteur(self, p):
        return arbre_abstrait.Entier(p.ENTIER)

    @_('IDENTIFIANT')
    def facteur(self, p):
        return arbre_abstrait.Identifiant(p.IDENTIFIANT)

    @_('IDENTIFIANT')
    def identifiant(self, p):
        return arbre_abstrait.Identifiant(p.IDENTIFIANT)

    @_('SI "(" expr ")" "{" prog "}"')
    def si(self, p):
        return arbre_abstrait.If(p.expr, p.prog, None)

    @_('si SINON SI "(" expr ")" "{" prog "}"')
    def si(self, p):
        p.si.add_elif(p.expr, p.prog)
        return p.si

    @_('si')
    def cond(self, p):
        return p.si

    @_('si SINON "{" prog "}"')
    def cond(self, p):
        p.si.add_else(p.prog)
        return p.si

    @_('TANTQUE "(" expr ")" "{" prog "}"')
    def instruction(self, p):
        return arbre_abstrait.While(p.expr, p.prog)

    @_('cond')
    def instruction(self, p):
        return p.cond

    @_('TYPE identifiant')
    def decl(self, p):
        return arbre_abstrait.Declaration(p.identifiant,
                                          arbre_abstrait.Type.from_str(p.TYPE),
                                          None)

    @_('decl ";"')
    def instruction(self, p):
        return p.decl

    @_('TYPE identifiant "=" expr ";"')
    def instruction(self, p):
        return arbre_abstrait.Declaration(p.identifiant,
                                          arbre_abstrait.Type.from_str(p.TYPE),
                                          p.expr)

    @_('identifiant "=" expr ";"')
    def instruction(self, p):
        return arbre_abstrait.Assignment(p.identifiant, p.expr)

    @_('RETOURNER expr ";"')
    def instruction(self, p):
        return arbre_abstrait.Return(p.expr)

    @_('function_arg "," expr')
    def function_arg(self, p):
        p.function_arg.append(p.expr)
        return p.function_arg

    @_('expr')
    def function_arg(self, p):
        return arbre_abstrait.FunctionArgs([p.expr])

    @_('decl')
    def function_decl_arg(self, p):
        return [p.decl]

    @_('function_decl_arg "," decl')
    def function_decl_arg(self, p):
        p.function_decl_arg.append(p.decl)
        return p.function_decl_arg

    @_('TYPE identifiant "(" function_decl_arg ")" "{" prog "}"')
    def function(self, p):
        return arbre_abstrait.Function(p.identifiant, p.function_decl_arg,
                                       p.prog,
                                       arbre_abstrait.Type.from_str(p.TYPE))

    @_('TYPE identifiant "(" ")" "{" prog "}"')
    def function(self, p):
        return arbre_abstrait.Function(p.identifiant, [], p.prog,
                                       arbre_abstrait.Type.from_str(p.TYPE))

    @_('function')
    def instruction(self, p):
        return p.function


def analyse_syntaxique(input: str) -> arbre_abstrait.Programme:
    lexer = FloLexer()
    parser = FloParser()
    return parser.parse(lexer.tokenize(input))
