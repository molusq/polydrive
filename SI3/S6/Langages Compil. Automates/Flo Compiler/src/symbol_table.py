from typing import Dict, List

from src.arbre_abstrait import Declaration, Identifiant, Programme, Type, Function


class Variable:

    def __init__(self, name: Identifiant, type: Type, offset: int, depth: int):
        self.name = name
        self.type = type
        self.offset = offset
        self.depth = depth


Symbol = Function | Variable


class SymbolTable:

    def __init__(self):
        self._symbols: Dict[str, Symbol] = {}

    @staticmethod
    def from_program(program: Programme):
        symbol_table = SymbolTable()
        for instruction in program.instructions:
            if isinstance(instruction, Function):
                symbol_table.add(instruction)
        return symbol_table

    def check_presence(self, sym: Identifiant):
        if not sym in self:
            raise Exception(f"Le symbole {sym} n'est pas défini")

    def add(self, sym: Symbol):
        if sym in self:
            raise Exception(f"Le symbole {sym.name} est déjà défini")
        self._symbols[sym.name.valeur] = sym

    def get(self, sym: Identifiant) -> Symbol:
        self.check_presence(sym)
        return self._symbols[sym.valeur]

    def remove(self, sym: Identifiant):
        self.check_presence(sym)
        del self._symbols[sym.valeur]

    def type(self, func: Identifiant) -> Type:
        self.check_presence(func)
        return self._symbols[func.valeur].return_type

    def args(self, func: Identifiant) -> List[Declaration]:
        self.check_presence(func)
        return self._symbols[func.valeur].args

    def memory_size(self, sym: Identifiant) -> int:
        self.check_presence(sym)
        func = self.get(sym)
        return 4 * len(func.args) + func.stack_size

    def memory_size_locals(self, sym: Identifiant) -> int:
        self.check_presence(sym)
        func = self.get(sym)
        return func.stack_size

    def next_address(self) -> int:
        return max([
            sym.offset
            for sym in self._symbols.values() if type(sym) == Variable
        ],
                   default=0) + 4

    def end_block(self, depth: int):
        in_depth = {
            k: v
            for k, v in self._symbols.items()
            if type(v) == Variable and v.depth == depth
        }
        for k in in_depth:
            del self._symbols[k]
        return in_depth

    def __contains__(self, sym: Symbol | Identifiant):
        if isinstance(sym, Identifiant):
            return sym.valeur in self._symbols

        return sym.name.valeur in self._symbols

    def __str__(self):
        return '\n'.join([str(symbol) for symbol in self._symbols.values()])