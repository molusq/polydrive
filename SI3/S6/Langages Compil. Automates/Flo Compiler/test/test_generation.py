import syrupy

from src.analyse_syntaxique import analyse_syntaxique
from src.generation_code import gen_entrypoint


def get_out(input: str, capsys):
    out = analyse_syntaxique(input)
    gen_entrypoint(out)
    captured = capsys.readouterr()
    return captured.out


def test_function(snapshot, capsys):
    input = """
    entier f() {
        retourner 3;
    }
    ecrire(f());
    """
    assert get_out(input, capsys) == snapshot


def test_invalid_call(snapshot, capsys):
    input = """
    entier f() {
        retourner 3;
    }
    ecrire(f(3));
    """

    try:
        get_out(input, capsys)
        assert False
    except Exception as e:
        assert e == snapshot


def test_invalid1(snapshot, capsys):
    input = """
    retourner 3;
    """

    try:
        get_out(input, capsys)
        assert False
    except Exception as e:
        assert e == snapshot


def test_invalid2(snapshot, capsys):
    input = """
        booleen f() {
        retourner 3;
    }
    """

    try:
        get_out(input, capsys)
        assert False
    except Exception as e:
        assert e == snapshot


def test_invalid3(snapshot, capsys):
    input = """
    entier f() {
        retourner Vrai;
    }
    """

    try:
        get_out(input, capsys)
        assert False
    except Exception as e:
        assert e == snapshot


def test_function_with_param(snapshot, capsys):
    input = """
    entier f(entier e) {
        ecrire(e);
        retourner 1;
    }
    """
    assert get_out(input, capsys) == snapshot


def test_affectation(snapshot, capsys):
    input = """
    entier f(entier e) {
        e = 7;
        ecrire(e);
        retourner 5;
    }
    f(5);
    """
    assert get_out(input, capsys) == snapshot


def test_if(snapshot, capsys):
    input = """
    entier f(entier e) {
        si (e > 5) {
            ecrire(e);
        }
        sinon {
            ecrire(5);
            ecrire(6);
        }
        retourner 1;
    }
    f(5);
    """
    assert get_out(input, capsys) == snapshot


def test_var_local1(snapshot, capsys):
    input = """
    entier f(entier e) {
        entier a = lire();
        retourner a + e;
    }
    f(5);
    """
    assert get_out(input, capsys) == snapshot


def test_var_local2(snapshot, capsys):
    input = """
    entier f(entier e) {
        si (e == 0) {
            entier a = 5;
        }
        si (e == 0) {
            entier a = 5;
        }
        si (e == 0) {
            entier a = 5;
        }
    }
    f(5);
    """
    assert get_out(input, capsys) == snapshot


def test_var_local3(snapshot, capsys):
    input = """
    entier f(entier e) {
        entier a = lire();
        si (a > 2) {
            entier b = a * a;
            si (b > 3 * a) {
                entier c = 25;
            }
            sinon {
                entier d = 26;
            }
        }
    }
    f(5);
    """
    assert get_out(input, capsys) == snapshot


def test_var_local_invalid1(snapshot, capsys):
    input = """
    entier f(booleen b) {
        si (b) {
            entier a = lire();
            ecrire(a);
        }
        ecrire(a + 5);
        retourner 1;
    }
    f(Vrai);
    """

    try:
        get_out(input, capsys)
        assert False
    except Exception as e:
        assert e == snapshot


def test_var_local_invalid2(snapshot, capsys):
    input = """
    entier f() {
        entier a;
    }
    entier g() {
        a = 5;
    }
    f();
    """

    try:
        get_out(input, capsys)
        assert False
    except Exception as e:
        assert e == snapshot


def test_var_local_invalid3(snapshot, capsys):
    input = """
    entier f(booleen b) {
        tantque (b) {
            entier rep = lire();
            si (rep > 10) {
                b = Faux;
            }
        }
        retourner rep;
    }
    f(Vrai);
    """

    try:
        get_out(input, capsys)
        assert False
    except Exception as e:
        assert e == snapshot


def test_var_global(snapshot, capsys):
    input = """
    entier f() {
        retourner 123;
    }
    entier a = f();
    """
    assert get_out(input, capsys) == snapshot
