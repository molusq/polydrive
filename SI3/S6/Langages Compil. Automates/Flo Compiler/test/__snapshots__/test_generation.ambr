# serializer version: 1
# name: test_affectation
  '''
  %include	"io.asm"
  section	.bss
  sinput:	resb	255	;reserve a 255 byte space in memory for the users input string
  v$a:	resd	1
  section	.text
  global _start
  _f:
  	push	7		
  	pop	eax		
  	mov	[ebp-4],	eax	
  	mov	eax,	[ebp-4]	
  	push	eax		
  	pop	eax		
  	call	iprintLF		
  	push	5		
  	pop	eax		
  	ret					 ; retourne à l'appel de la fonction
  _start:
  __main:
  	push	ebp		
  	sub	esp,	0	
  	push	5		
  	call	_f		
  	add	esp,	4	
  	pop	ebp		
  	push	eax		
  	pop	eax		
  	mov	eax,	1			 ; 1 est le code de SYS_EXIT
  	int	0x80				 ; exit
  
  '''
# ---
# name: test_function
  '''
  %include	"io.asm"
  section	.bss
  sinput:	resb	255	;reserve a 255 byte space in memory for the users input string
  v$a:	resd	1
  section	.text
  global _start
  _f:
  	push	3		
  	pop	eax		
  	ret					 ; retourne à l'appel de la fonction
  _start:
  __main:
  	push	ebp		
  	sub	esp,	0	
  	call	_f		
  	add	esp,	0	
  	pop	ebp		
  	push	eax		
  	pop	eax		
  	call	iprintLF		
  	mov	eax,	1			 ; 1 est le code de SYS_EXIT
  	int	0x80				 ; exit
  
  '''
# ---
# name: test_function_with_param
  '''
  	mov	eax,	[ebp-4]	
  	push	eax		
  	pop	eax		
  	call	iprintLF		
  	push	1		
  	pop	eax		
  	ret					 ; retourne à l'appel de la fonction
  _start:
  __main:
  	mov	eax,	1			 ; 1 est le code de SYS_EXIT
  	int	0x80				 ; exit
  
  '''
# ---
# name: test_if
  '''
  %include	"io.asm"
  section	.bss
  sinput:	resb	255	;reserve a 255 byte space in memory for the users input string
  v$a:	resd	1
  section	.text
  global _start
  _f:
  	mov	eax,	[ebp-4]	
  	push	eax		
  	push	5		
  	pop	ebx				 ; dépile la seconde operande dans ebx
  	pop	eax				 ; dépile la permière operande dans eax
  	cmp	eax,	ebx			 ; compare eax et ebx
  	setg	al				 ; met al à 1 si eax > ebx
  	push	eax				 ; empile le résultat
  	pop	eax		
  	cmp	eax,	0	
  	je	e2		
  	mov	eax,	[ebp-4]	
  	push	eax		
  	pop	eax		
  	call	iprintLF		
  	jmp	e3		
  	e2:			
  	push	5		
  	pop	eax		
  	call	iprintLF		
  	push	6		
  	pop	eax		
  	call	iprintLF		
  	e3:			
  	push	1		
  	pop	eax		
  	ret					 ; retourne à l'appel de la fonction
  _start:
  __main:
  	push	ebp		
  	sub	esp,	0	
  	push	5		
  	call	_f		
  	add	esp,	4	
  	pop	ebp		
  	push	eax		
  	pop	eax		
  	mov	eax,	1			 ; 1 est le code de SYS_EXIT
  	int	0x80				 ; exit
  
  '''
# ---
# name: test_invalid1
  Exception('return de type différent de la fonction')
# ---
# name: test_invalid2
  Exception('return de type différent de la fonction')
# ---
# name: test_invalid3
  Exception('return de type différent de la fonction')
# ---
# name: test_invalid_call
  IndexError('list index out of range')
# ---
# name: test_var_global
  '''
  	push	123		
  	pop	eax		
  	ret					 ; retourne à l'appel de la fonction
  _start:
  __main:
  	push	ebp		
  	sub	esp,	0	
  	call	_f		
  	add	esp,	0	
  	pop	ebp		
  	push	eax		
  	pop	eax		
  	mov	[ebp-4],	eax	
  	mov	eax,	1			 ; 1 est le code de SYS_EXIT
  	int	0x80				 ; exit
  
  '''
# ---
# name: test_var_local1
  '''
  %include	"io.asm"
  section	.bss
  sinput:	resb	255	;reserve a 255 byte space in memory for the users input string
  v$a:	resd	1
  section	.text
  global _start
  _f:
  	mov	eax,	sinput	
  	call	readline		
  	call	atoi		
  	push	eax		
  	pop	eax		
  	mov	[ebp-8],	eax	
  	mov	eax,	[ebp-8]	
  	push	eax		
  	mov	eax,	[ebp-4]	
  	push	eax		
  	pop	ebx				 ; dépile la seconde operande dans ebx
  	pop	eax				 ; dépile la permière operande dans eax
  	add	eax,	ebx			 ; effectue l'opération eax+ebx et met le résultat dans eax
  	push	eax				 ; empile le résultat
  	pop	eax		
  	ret					 ; retourne à l'appel de la fonction
  _start:
  __main:
  	push	ebp		
  	sub	esp,	4	
  	push	5		
  	call	_f		
  	add	esp,	8	
  	pop	ebp		
  	push	eax		
  	pop	eax		
  	mov	eax,	1			 ; 1 est le code de SYS_EXIT
  	int	0x80				 ; exit
  
  '''
# ---
# name: test_var_local2
  '''
  %include	"io.asm"
  section	.bss
  sinput:	resb	255	;reserve a 255 byte space in memory for the users input string
  v$a:	resd	1
  section	.text
  global _start
  _f:
  	mov	eax,	[ebp-4]	
  	push	eax		
  	push	0		
  	pop	ebx				 ; dépile la seconde operande dans ebx
  	pop	eax				 ; dépile la permière operande dans eax
  	cmp	eax,	ebx			 ; compare eax et ebx
  	sete	al				 ; met al à 1 si eax == ebx
  	push	eax				 ; empile le résultat
  	pop	eax		
  	cmp	eax,	0	
  	je	e7		
  	push	5		
  	pop	eax		
  	mov	[ebp-8],	eax	
  	e7:			
  	mov	eax,	[ebp-4]	
  	push	eax		
  	push	0		
  	pop	ebx				 ; dépile la seconde operande dans ebx
  	pop	eax				 ; dépile la permière operande dans eax
  	cmp	eax,	ebx			 ; compare eax et ebx
  	sete	al				 ; met al à 1 si eax == ebx
  	push	eax				 ; empile le résultat
  	pop	eax		
  	cmp	eax,	0	
  	je	e8		
  	push	5		
  	pop	eax		
  	mov	[ebp-8],	eax	
  	e8:			
  	mov	eax,	[ebp-4]	
  	push	eax		
  	push	0		
  	pop	ebx				 ; dépile la seconde operande dans ebx
  	pop	eax				 ; dépile la permière operande dans eax
  	cmp	eax,	ebx			 ; compare eax et ebx
  	sete	al				 ; met al à 1 si eax == ebx
  	push	eax				 ; empile le résultat
  	pop	eax		
  	cmp	eax,	0	
  	je	e9		
  	push	5		
  	pop	eax		
  	mov	[ebp-8],	eax	
  	e9:			
  _start:
  __main:
  	push	ebp		
  	sub	esp,	12	
  	push	5		
  	call	_f		
  	add	esp,	16	
  	pop	ebp		
  	push	eax		
  	pop	eax		
  	mov	eax,	1			 ; 1 est le code de SYS_EXIT
  	int	0x80				 ; exit
  
  '''
# ---
# name: test_var_local3
  '''
  %include	"io.asm"
  section	.bss
  sinput:	resb	255	;reserve a 255 byte space in memory for the users input string
  v$a:	resd	1
  section	.text
  global _start
  _f:
  	mov	eax,	sinput	
  	call	readline		
  	call	atoi		
  	push	eax		
  	pop	eax		
  	mov	[ebp-8],	eax	
  	mov	eax,	[ebp-8]	
  	push	eax		
  	push	2		
  	pop	ebx				 ; dépile la seconde operande dans ebx
  	pop	eax				 ; dépile la permière operande dans eax
  	cmp	eax,	ebx			 ; compare eax et ebx
  	setg	al				 ; met al à 1 si eax > ebx
  	push	eax				 ; empile le résultat
  	pop	eax		
  	cmp	eax,	0	
  	je	e13		
  	mov	eax,	[ebp-8]	
  	push	eax		
  	mov	eax,	[ebp-8]	
  	push	eax		
  	pop	ebx				 ; dépile la seconde operande dans ebx
  	pop	eax				 ; dépile la permière operande dans eax
  	imul	ebx				 ; effectue l'opération eax*ebx et met le résultat dans eax
  	push	eax				 ; empile le résultat
  	pop	eax		
  	mov	[ebp-12],	eax	
  	mov	eax,	[ebp-12]	
  	push	eax		
  	push	3		
  	mov	eax,	[ebp-8]	
  	push	eax		
  	pop	ebx				 ; dépile la seconde operande dans ebx
  	pop	eax				 ; dépile la permière operande dans eax
  	imul	ebx				 ; effectue l'opération eax*ebx et met le résultat dans eax
  	push	eax				 ; empile le résultat
  	pop	ebx				 ; dépile la seconde operande dans ebx
  	pop	eax				 ; dépile la permière operande dans eax
  	cmp	eax,	ebx			 ; compare eax et ebx
  	setg	al				 ; met al à 1 si eax > ebx
  	push	eax				 ; empile le résultat
  	pop	eax		
  	cmp	eax,	0	
  	je	e14		
  	push	25		
  	pop	eax		
  	mov	[ebp-16],	eax	
  	jmp	e15		
  	e14:			
  	push	26		
  	pop	eax		
  	mov	[ebp-16],	eax	
  	e15:			
  	e13:			
  _start:
  __main:
  	push	ebp		
  	sub	esp,	16	
  	push	5		
  	call	_f		
  	add	esp,	20	
  	pop	ebp		
  	push	eax		
  	pop	eax		
  	mov	eax,	1			 ; 1 est le code de SYS_EXIT
  	int	0x80				 ; exit
  
  '''
# ---
# name: test_var_local_invalid1
  Exception('Le symbole {\n  "identifiant": "a"\n} n\'est pas défini')
# ---
# name: test_var_local_invalid2
  Exception('Le symbole {\n  "identifiant": "a"\n} n\'est pas défini')
# ---
# name: test_var_local_invalid3
  Exception('Le symbole {\n  "identifiant": "rep"\n} n\'est pas défini')
# ---
